chips/cc26x2/src/rfc.rs

```diff
@@ -0,0 +1,580 @@
+
+        self.handle_cpe_interrupts();
+        self.handle_hw_interrupts();
+
+        // Initialize radio module
+        self.send_direct(&cmd::DirectCommand::new(cmd::RFC_CMD0, 0x10 | 0x40))
+            .ok()
+            .expect("Could not initialize radio module");
+
+        // Request bus
+        self.send_direct(&cmd::DirectCommand::new(cmd::RFC_BUS_REQUEST, 1))
+            .ok()
+            .expect("Could not request bus for radio module");
+
+        // Ping radio module
+        self.send_direct(&cmd::DirectCommand::new(cmd::RFC_PING, 0))
+            .ok()
+            .expect("Coudl not ping radio module");
+    }
+
+    pub fn disable(&self) {
+        self.send_direct(&cmd::DirectCommand::new(cmd::RFC_STOP, 0))
+            .ok()
+            .expect("Could not send stop cmd to radio module");
+
+        self.disable_cpe_interrupts();
+        self.disable_hw_interrupts();
+
+        let fs_powerdown: cmd::CmdFSPowerdown = cmd::CmdFSPowerdown::new();
+
+        self.send(&fs_powerdown)
+            .and_then(|_| self.wait(&fs_powerdown))
+            .ok()
+            .expect("Could not power down frequency synthesizer for radio module");
+
+        self.stop_rat();
+
+        // Add disable power domain and clocks
+
+        self.mode.set(None);
+    }
+
+    pub fn setup(&self, reg_override: u32, tx_power: u16) {
+        let mode = self.mode.get().expect("No RF mode selected, cannot setup");
+        let radio_setup = cmd::CmdRadioSetup::new(reg_override, mode as u8, tx_power);
+
+        self.send(&radio_setup)
+            .and_then(|_| self.wait(&radio_setup))
+            .ok()
+            .expect("Could not enable NonProp mode in radio module");
+    }
+
+    pub fn current_mode(&self) -> Option<RfcMode> {
+        self.mode.get()
+    }
+
+    pub fn set_mode(&self, mode: RfcMode) {
+        let rf_mode = match mode {
+            RfcMode::NONPROP => 0x00,
+            _ => panic!("Only HAL mode supported"),
+        };
+
+        prcm::rf_mode_sel(rf_mode);
+
+        self.mode.set(Some(mode))
+    }
+
+    fn post_cmdr(&self, rf_command: u32) -> RfcResult {
+        let dbell_regs = RFC_DBELL_BASE;
+
+        if !prcm::Power::is_enabled(prcm::PowerDomain::RFC) {
+            panic!("RFC power domain is off");
+        }
+
+        dbell_regs.cmdr.set(rf_command);
+
+        let mut timeout: u32 = 0;
+        let mut status = 0;
+
+        const MAX_TIMEOUT: u32 = 0x2FFFFFF;
+        while timeout < MAX_TIMEOUT {
+            status = self.cmdsta();
+            if (status & 0xFF) == 0x01 {
+                return Ok(());
+            }
+
+            timeout += 1;
+        }
+
+        return Err(status);
+    }
+
+    fn wait_cmdr(&self, rf_command: u32) -> RfcResult {
+        let command_op: &cmd::CmdCommon = unsafe { &*(rf_command as *const cmd::CmdCommon) };
+
+        let mut timeout: u32 = 0;
+        let mut status = 0;
+        const MAX_TIMEOUT: u32 = 0x2FFFFFF;
+        while timeout < MAX_TIMEOUT {
```

Comment:
Blocking on I/O operations to completely like this is a problem in Tock, since the kernel is not internally pre-emtive. In this case, because the timeout is so high, it looks like we could be waiting for **_second_** at 48Mhz... which means the kernel is probably gonna miss important interrupts.

Not having looked that deeply into the radio portion of the datasheet, is there a way to do this in an interrupt driven fashion instead of spinning?

Reply:
Looks like the `COMMAND_DONE` and `LAST_COMMAND_DONE` interrupts from the radio CPU might do what we want here.

Reply:
removed timeout. added in comments about building this out going forward.